crates/red_knot_python_semantic/resources/mdtest/metaclass.md

```diff
@@ -0,0 +1,191 @@
+```
+
+## Invalid metaclass
+
+If a class is a subclass of a class with a custom metaclass, then the subclass will also have that
+metaclass.
+
+```py
+class A: ...
+class B(metaclass=A): ...
+
+# TODO: emit a diagnostic for the invalid metaclass
+reveal_type(B.__class__)  # revealed: Literal[A]
+```
+
+## Linear inheritance
+
+If a class is a subclass of a class with a custom metaclass, then the subclass will also have that
+metaclass.
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+class C(B): ...
+
+reveal_type(C.__class__)  # revealed: Literal[A]
+```
+
+## Conflict (1)
+
+The metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its
+bases. ("Strict subclass" is a synonym for "proper subclass"; a non-strict subclass can be a
+subclass or the class itself.)
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+class C(type): ...
+class D(metaclass=C): ...
+
+# error: [conflicting-metaclass] "The metaclass of a derived class (`E`) must be a subclass of the metaclasses of all its bases, but `C` and `A` have no subclass relationship"
+class E(D, B): ...
+
+reveal_type(E.__class__)  # revealed: Unknown
+```
+
+## Conflict (2)
+
+The metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its
+bases. ("Strict subclass" is a synonym for "proper subclass"; a non-strict subclass can be a
+subclass or the class itself.)
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+class C(type): ...
+
+# error: [conflicting-metaclass] "The metaclass of a derived class (`E`) must be a subclass of the metaclasses of all its bases, but `C` and `A` have no subclass relationship"
+class E(B, metaclass=C): ...
+
+reveal_type(E.__class__)  # revealed: Unknown
+```
+
+## Common metaclass
+
+A class has two explicit bases, both of which have the same metaclass.
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+class C(metaclass=A): ...
+class D(B, C): ...
+
+reveal_type(D.__class__)  # revealed: Literal[A]
+```
+
+## Metaclass metaclass
+
+A class has an explicit base with a custom metaclass. That metaclass itself has a custom metaclass.
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+class C(B): ...
+class D(metaclass=C): ...
+class E(D): ...
+
+reveal_type(E.__class__)  # revealed: Literal[C]
+```
+
+## Diamond inheritance
+
+```py
+class M(type): ...
+class M1(M): ...
+class M2(M): ...
+class M12(M1, M2): ...
+class A(metaclass=M1): ...
+class B(metaclass=M2): ...
+class C(metaclass=M12): ...
+
+# error: [conflicting-metaclass] "The metaclass of a derived class (`D`) must be a subclass of the metaclasses of all its bases, but `M1` and `M2` have no subclass relationship"
+class D(A, B, C): ...
+
+reveal_type(D.__class__)  # revealed: Unknown
+```
+
+## Unknown
+
+```py
+from nonexistent_module import UnknownClass  # error: [unresolved-import]
+
+class C(UnknownClass): ...
+
+reveal_type(C.__class__)  # revealed: Literal[type]
```

Comment:
@carljm -- In your example, this was `# revealed: type[type] & Unknown` -- am I missing an `Unknown` here?

Reply:
Yes, but I think you should just add a TODO for it, since you also need the `type[]` type, which hasn't landed yet (and the distinction here may not matter much practically, so we can defer the TODO until/unless we find a case where it does matter)
```suggestion
# TODO should technically be `type[type] & Unknown`
reveal_type(C.__class__)  # revealed: Literal[type]
```